### ISSUE TITLE
NUnit extensibility: tests batching support

### ISSUE THREAD

**Issue by halex2005** [nunit/nunit#2206]
In NUnit 2 I have had implemented `IAddin`, which installs custom implemented `ITestDecorator`.
In implementation of installed `ITestDecorator` I was able to dynamically decide if test fixture need to run or should be skipped. This was perfect place to make tests batching: to run each test batch on different teamcity agents "in parallel".

In NUnit 3 the extensibility model was significantly changed. Would you tell me how I can achieve similar functionality in NUnit 3? 

One solution I found is complicated. I can:

- dump all the tests to file with `--explore` switch, 
- process the file and keep only tests for single batch to run, 
- run console runner with `--testlist` switch to actually run tests

Would you offer a better solution?

**Comment by CharliePoole** [comment]
I'd need to know what criterion you want to use to run or not run the test. How did your addin do it?

**Comment by halex2005** [comment]
For example, I have four teamcity configurations, each has two parameters:

- current batch number
- overall batches count

In implementation of `ITestDecorator` I can compute hash for test fixture and divide it by module of "overall batches count". Then I compare result with "current batch number", if they are equal test fixture will run all the tests. Otherwise, tests will be skipped.

Using hash is the simplest algorithm. But tests distribution is not uniform.

To minimize nonuniformness of tests distribution, now I accumulate tests count passed for batch in array. Next batch index is selected from this array, I search index with minimal value for next batch index.

I think you got an idea.

**Comment by CharliePoole** [comment]
How does your test decorator know to which tests it should be applied. Do you implement a special attribute? Or do you just decorate every fixture?

**Comment by halex2005** [comment]
No, there are no any special attributes, the test decorator is applied to all tests in current run

**Comment by jnm2** [comment]
@halex2005 I wonder if the new parallelization support in 3.7 is able to replace your method of manually batching in order to parallelize?

**Comment by CharliePoole** [comment]
@jnm2 He wants to parallelize parts of the same assembly onto different machines at a fixture level, if I understand it. So if running in the same machine would serve him, he could have used NUnit's parallelization even before 3.7. But we don't support multiple machines yet, so the only way to do it is to have multiple runs and tell each of them to only run certain tests. @halex2005 Please correct me if I didn't understand you correctly.

@halex2005 My first thought would be to use the `IWrapCommand` interface to insert a decorator command in the chain of commands executed when running the test. However, that requires you to define an an attribute and apply it to all your fixtures - not very practical.

The only way we support right now for running something on __every__ test, without adding an attribute to each one, is through an ActionAttribute. To be honest, I'm not entirely sure if this will work, but you would do it like this:

1. Define an action attribute, derived from TestActionAttribute.

2. Apply the attribute to the assembly, specifying `ActionTargets.Suite`.

3. Override BeforeTest: __IFF__ this test is a test fixture, apply your algorithm and mark it as skipped if you don't want to run it.

The part I am not sure of is whether it's too late to skip the test once we already executing action attributes on it. I don't have time to test this right now, so I'll leave it to you. Please let us know what you find out. If it's negative, we can search for another way.

**Comment by halex2005** [comment]
@jnm2, no, I think it is impossible.
We have many long running web tests (about 1000). Web tests cannot be parallelized, they are constrained to be single-threaded.

**Comment by halex2005** [comment]
@CharliePoole, you understood me correctly.

Thanks, I will try to use `TestActionAttribute` as you advised. Hope it will works))

To be clear, in Nunit v2 I used [this test decorators extension point](https://nunit.org/index.php?p=testDecorators&r=2.6.4). To run test I just return test passed in arguments, to skip test I set property of `test.RunState = RunState.Skipped`.

**Comment by CharliePoole** [comment]
@halex2005

That's the part I'm not sure of. You should do the same thing with RunState and see what happens. It may not do the job because the test is already running when the BeforeTest method is called. If not, you want to simulate dynamically skipping a test as Assert.Ignore does, but you don't want to use that because it would be reported as ignored. Instead, you want to silently skip the test.

What you hope to end up with is a result of ResultState.Skipped or perhaps ResultState.Inconclusive, with the test not actually being executed after you return from BeforeTest. Since you are probably not working with the NUnit source, you can't step through our code, but you can put a breakpoint in the very next thing in your code, which would be executed after the BeforeTest. That's probably a OneTimeSetUp in a fixture.

**Comment by halex2005** [comment]
@CharliePoole, unfortunately, it does not work.

First, `ITest.RunState` has no setter, only getter.
Second, you right, [it seems like it's too late to modify test state in this attribute](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/Commands/BeforeTestCommand.cs#L48), there is no checks if RunState was changed from `Runnable` to other value here.

**Comment by halex2005** [comment]
@CharliePoole , is there any chance that the team will consider to add extensibility point [like nunit v2 test decorators](https://nunit.org/index.php?p=testDecorators&r=2.6.4) in near future? 
Can I help with PR? What is your vision about this feature?

**Comment by halex2005** [comment]
Or

- I can try to cast `ITest` to `Test`. It has setter for `RunState` property
- We need PR with added logic in commands: need to check `RunState == RunState.Runnable` before executing test

Would you confirm this will works? If yes, I'll propose PR with changes needed.

**Comment by CharliePoole** [comment]
I looked at the code more closely and Action Attributes would not work anyway. The Before Test is not called until the fixture has already been constructed and the one time setup has run. You want to prevent running much earlier.

Our "equivalent" to TestDecorator actually is an execution-time decorator. That is, it runs the test and returns a modified result. That is of course not helpful to you.

Essentially, what you need is the functional equivalent of IApplyToTest, the need to apply an attribute... something you could define, which would be called on every test. It would be a new feature. 

@nunit/framework-team What do you think about this idea?

**Comment by ChrisMaddock** [comment]
@CharliePoole - Sorry, I'm not familiar with NUnit 2 addins at all, which is making this a little hard to follow. Can you explain the difference between what you're proposing, and a standard custom attribute?

**Comment by CharliePoole** [comment]
@ChrisMaddock Let's see how brief I can be!

NUnit V2 Addins are actually more like engine extensions that V3 custom attributes, even though an addin frequently had a custom attribute associated with it. That's because V2 attributes were basically markers - as .NET attributes usually are - and the code to act upon them was entirely separate. As you know, V3 attributes are active - they have lots of code in them.

A V2 addin, once installed, would be queried whenever the particular extension point was reached. The one used by @halex2005 was in nunit.core and extended the way we recognize and build test fixtures. Everytime we were building a fixture from a class, it would intervene and possibly change how the fixture was built. In this case, it changed the runstate of the fixture.

The key thing here is that because @halex2005 wanted his addin to apply to __every__ fixture created, he didn't define an attribute at all. Our architecture does not have a way to do the same thing unless a custom attribute is used. This is a hole in our functionality, which I think we should consider plugging.

Ideally, we should be able to implement globally just about any of the interfaces that we permit on custom attributes and have them apply everywhere. So, for example, when we are constructing tests and calling into all the attributes that implement `IApplyToTest` we should check to see if there are any global `IApplyToTest` calls that must be made.

It's an interesting problem.

**Comment by ChrisMaddock** [comment]
Thanks Charlie. Sounds sensible enough to me.

**Comment by jnm2** [comment]
I'm okay with it.

**Comment by jnm2** [comment]
@CharliePoole
> Our architecture does not have a way to do the same thing unless a custom attribute is used. This is a hole in our functionality, which I think we should consider plugging.
>
> Ideally, we should be able to implement globally just about any of the interfaces that we permit on custom attributes and have them apply everywhere. So, for example, when we are constructing tests and calling into all the attributes that implement IApplyToTest we should check to see if there are any global IApplyToTest calls that must be made.

This still needs design, right? What if we did it this way?

```c#
[ApplyGlobally]
public class MyExtensions : IApplyToTest, ...
{
    // ...
}
```

**Comment by CharliePoole** [comment]
@jnm2 It would be inconsistent with our current design, since we don't use attributes on attributes right now. I'd be more in favor of an interface that let an attribute decide what it applied on.

Are we using this issue to plan out the feature? It might need a new title in that case.

In terms of process, this is a discussion item. I made it one because I thought we should decide as a team whether to do accept or reject it. We should decide that first.

**Comment by jnm2** [comment]
@CharliePoole I was imagining that `MyExtensions` would not be an attribute. It shouldn't be applied to classes because NUnit will see the `[ApplyGlobally]` and manage that.

> Are we using this issue to plan out the feature? It might need a new title in that case.

I don't know, either sounds good.

> In terms of process, this is a discussion item. I made it one because I thought we should decide as a team whether to do accept or reject it. We should decide that first.

I took your recommendation, Chris's support and my support to mean that we've decided to accept it.In other threads we didn't wait for more than one or two to chime in and it's been 28 days.
